cumulusci/tasks/robotframework/robotperflistener.py

```diff
@@ -0,0 +1,142 @@
+import os
+import json
+from statistics import mean
+import time
+
+
+class TestMetrics:
+    """Manage both REST API (predefined) metrics and also custom metrics"""
+
+    def __init__(self):
+        self.api_metrics = []
+        self.custom_metrics = {}
+
+
+class RobotPerfListener:
+    """A robot listener for performance metrics from REST APIs and other code"""
+
+    ROBOT_LISTENER_API_VERSION = 2
+
+    def __init__(self, task, verbosity):
+        self._metrics = {}
+        self.task = task
+        self.verbosity = verbosity
+
+    def start_suite(self, name, attributes=None):
+        self._current_suite = {}
+        self._metrics[name] = self._current_suite
+
+    def end_suite(self, name, attributes=None):
+        self._current_suite = None
+
+    def start_test(self, name, attributes=None):
+        self._current_test = TestMetrics()
+
+    def end_test(self, name, attributes=None):
+        totals = summarize(self._current_test.api_metrics)
+        totals.update(
+            {
+                name: metric.value
+                for name, metric in self._current_test.custom_metrics.items()
+            }
+        )
+
+        test_result = {"totals": dict(sorted(totals.items()))}
+
+        if self.verbosity >= 1:
+            test_result["traces"] = self._current_test.api_metrics
+
+        self._current_suite[name] = test_result
+        self._current_test = None
+
+    def close(self):
+        filename = os.path.join(self.outputdir, "perf.json")
+        with open(filename, "w") as f:
+            json.dump(self._metrics, f, indent=2)
+
+        # cleanup so we won't be called again and also remove GC loop
+        self.task.robot_perf_listener = None
+        self.task = None
+
+    def report(self, metrics):
```

Comment:
I spent several minutes trying to figure out what was going on here. I was thinking this was a typo and meant to be the `report_file` method described in the API. I eventually realized it's an additional method called outside of robot. 

You might want to consider somehow segregating this method, or renaming it (perhaps with _ prefix) and the other non-listener-api methods so it's a bit more clear which are api methods and which are not. For example, simply moving it after `output_file` and adding a comment saying something like "these methods are not used by the listener interface, but are used by other parts of our code" or something similar would make the code a bit easier to grok. 

Another option might be to move the non-listener methods into a class mixin so that it's clear you are adding additional functions to a standard interface.

Reply:
Okay I'll look at something like that.

Reply:
167a3215527ac69b89236be0445247b434bcfce5